packages/grid/x-data-grid/src/utils/fastObjectShallowCompare.ts

```diff
@@ -0,0 +1,34 @@
+const is = Object.is;
+
+export function fastObjectShallowCompare<T extends Record<string, any> | null>(a: T, b: T) {
+  if (a === b) {
+    return true;
+  }
+  if (!(a instanceof Object) || !(b instanceof Object)) {
+    return false;
+  }
+
+  let aLength = 0;
+  let bLength = 0;
+
+  /* eslint-disable no-restricted-syntax */
+  /* eslint-disable guard-for-in */
+  for (const key in a) {
+    aLength += 1;
+
+    if (!(key in b)) {
+      return false;
+    }
+    if (!is(a[key], b[key])) {
+      return false;
+    }
+  }
+
+  for (const _ in b) {
+    bLength += 1;
+  }
+  /* eslint-enable no-restricted-syntax */
+  /* eslint-enable guard-for-in */
+
+  return aLength === bLength;
+}
```

Comment:
I found another optimization, instead of validating in the second loop that `b` keys are in `a`, we just compute the length of `b` keys. And because we already check in the `a` loop that all `a` keys are in `b`, we can simply compare the keys length
. We can't bail out of the `b` loop, but arithmetic ops are so cheap compared to key checks that it's worth it.

Benchmark: https://jsben.ch/aZKxv

| | Chrome | Firefox |
| --- | --- | --- |
| a & b are equal | ![Screenshot from 2023-06-02 07-47-08](https://github.com/mui/mui-x/assets/1423607/1bc442ff-0644-4235-857b-5b49b4800944) | ![Screenshot from 2023-06-02 07-50-14](https://github.com/mui/mui-x/assets/1423607/6de5451b-a4f2-4b7c-9db2-8d121a8f3a29) |
| a & b's first key differs | ![Screenshot from 2023-06-02 07-49-16](https://github.com/mui/mui-x/assets/1423607/0915ace5-0baf-45cc-8cf5-1386ca4fdf9f) | ![Screenshot from 2023-06-02 07-50-45](https://github.com/mui/mui-x/assets/1423607/426031b3-b6d5-4651-ad70-8cd39291862c) |

Interestingly, `fastOSC improved` beats `fastOSC` even if `a` and `b` differ on their first key. But when the objects are equal, `fastOSC improved` is by far superior to the alternatives.

Reply:
Nice! Totally makes sense, since we don't need to know which keys of `b` are missing in `a` 🎉

Reply:
^ This is just me being fascinated by performance improvements, but in practice object shallow equality checks are already so fast that this doesn't matter much. No need to pay much attention to this.

Reply:
Got another small improvement by inverting the order of the two if branches in the `a` loop:

```typescript
    if (!is(a[key], b[key])) {
      return false;
    }
    if (!(key in b)) {
      return false;
    }
```

But it only holds for V8. On SpiderMonkey, inverting the branches doesn't yield any significant change.

https://jsben.ch/obBsU

<img src="https://github.com/mui/mui-x/assets/1423607/6708428f-e922-47e7-98d2-777540ef3b2a" width="400" />